File: databricks_notebooks/python/mlflow/02_Read_SparkBundle.py
# Databricks notebook source
# MAGIC %md ## Read MLeap bundle as SparkBundle from MLflow

# COMMAND ----------

# MAGIC %md ### Setup

# COMMAND ----------

# MAGIC %run ../Common

# COMMAND ----------

# MAGIC %md ### Read data

# COMMAND ----------

data = spark.read.csv(data_path, header="true", inferSchema="true")
(trainingData, testData) = data.randomSplit([0.7, 0.3], 2019)

# COMMAND ----------

# MAGIC %md ### Get MLflow run

# COMMAND ----------

import mlflow
client = mlflow.tracking.MlflowClient()
experiment_id = client.get_experiment_by_name(experiment_name_python).experiment_id

# COMMAND ----------

run = client.search_runs(experiment_id, order_by=["attributes.start_time desc"], max_results=1)[0]
run

# COMMAND ----------

run.info.artifact_uri

# COMMAND ----------

# MAGIC %md ### Read MLeap bundle

# COMMAND ----------

# MAGIC %run ./MLeap_Utils

# COMMAND ----------

model = load_model_as_spark_bundle(run, "mleap-model/mleap/model")

# COMMAND ----------

# MAGIC %md ### Run predictions

# COMMAND ----------

predictions = model.transform(data)

# COMMAND ----------

display(predictions.select(colPrediction, colLabel, colFeatures))

File: databricks_notebooks/scala/mlflow/02_Read_SparkBundle.scala
// Databricks notebook source
// MAGIC %md ## Read MLeap bundle as a SparkBundle from MLflow
// MAGIC * Widget `bundle` - which language was used to generate bundle.

// COMMAND ----------

// MAGIC %md ### Setup

// COMMAND ----------

// MAGIC %run ../Common

// COMMAND ----------

// MAGIC %run ../SparkBundleUtils

// COMMAND ----------

// MAGIC %run ./MLflowUtils

// COMMAND ----------

// MAGIC %md ### Get the last MLflow run ID

// COMMAND ----------

import org.mlflow.tracking.{MlflowClient,MlflowContext,MlflowClientVersion}
val client = new MlflowClient()
println("MLflow version: "+MlflowClientVersion.getClientVersion())

// COMMAND ----------

val bundle = createBundleWidget()
val experimentName = getExperimentNameForBundle(bundle) 
val runId = MlflowUtils.getLastRunId(client, experimentName)

// COMMAND ----------

// MAGIC %md ### Read data

// COMMAND ----------

val data = spark.read.format("csv").option("header", "true").option("inferSchema", "true").load(dataPathDbfs)

// COMMAND ----------

// MAGIC %md ### Read bundle as SparkBundle

// COMMAND ----------

val modelPath = client.downloadArtifacts(runId,"mleap-model/mleap/model").getAbsolutePath

// COMMAND ----------

val bundlePath = s"file:${modelPath}"
val model = SparkBundleUtils.readModel(bundlePath)

// COMMAND ----------

// MAGIC %md ### Run predictions

// COMMAND ----------

val predictions = model.transform(data)

// COMMAND ----------

// MAGIC %md ### Show predictions

// COMMAND ----------

println(f"Prediction count: ${predictions.count}")
val sum = predictions.agg(Map("prediction"->"sum")).take(1)(0).getDouble(0)
println(f"Prediction sum:   ${sum}%.3f")

// COMMAND ----------

import org.apache.spark.sql.functions._
predictions.groupBy("prediction").count().sort(desc("count")).limit(10).show

// COMMAND ----------

display(predictions.select("prediction","quality","features").sort("prediction","quality","features"))